url8d1df200a8_gilleain_signatures_tgz-pJ8-signature_chemistry_LargeMoleculeTestJ8/src/signature/DAG.java

```diff
@@ -0,0 +1,682 @@
+            } else {
+                arcs.add(arc);
+                getOccurences(occurences, child, node, arcs);
+            }
+        }
+    }
+	
+	public List<InvariantInt> getInvariantPairs(int[] parents) {
+	    List<InvariantInt> pairs = new ArrayList<InvariantInt>();
+	    for (int i = 0; i < this.vertexCount; i++) {
+	        if (invariants.getColor(i) == -1
+	                && parents[i] >= 2) {
+	            pairs.add(
+	                    new InvariantInt(
+	                            invariants.getVertexInvariant(i), i));
+	        }
+	    }
+	    Collections.sort(pairs);
+	    return pairs;
+	}
+	
+	public int colorFor(int vertexIndex) {
+		return this.invariants.getColor(vertexIndex);
+	}
+	
+	public void accept(DAGVisitor visitor) {
+	    this.getRoot().accept(visitor);
+	}
+
+	public void addLayer(List<Node> layer) {
+		this.layers.add(layer);
+	}
+	
+	public List<Integer> createOrbit(int[] parents) {
+	    
+	    // get the orbits
+	    Map<Integer, List<Integer>> orbits = 
+	        new HashMap<Integer, List<Integer>>();
+	    for (int j = 0; j < vertexCount; j++) {
+	        if (parents[j] >= 2) {
+	            int invariant = invariants.getVertexInvariant(j);
+	            List<Integer> orbit;
+	            if (orbits.containsKey(invariant)) {
+	                orbit = orbits.get(invariant);
+	            } else {
+	                orbit = new ArrayList<Integer>();
+	                orbits.put(invariant, orbit);
+	            }
+	            orbit.add(j);
+	        }
+	    }
+	    
+//	    System.out.println("Orbits " + orbits);
+	    
+	    // find the largest orbit
+	    if (orbits.isEmpty()) {
+	        return new ArrayList<Integer>();
+	    } else {
+	        List<Integer> maxOrbit = null;
+	        List<Integer> invariants = new ArrayList<Integer>(orbits.keySet());
+	        Collections.sort(invariants);
+	        
+	        for (int invariant : invariants) {
+	            List<Integer> orbit = orbits.get(invariant);
+	            if (maxOrbit == null || orbit.size() > maxOrbit.size()) { 
+	                maxOrbit = orbit;
+	            }
+	        }
+	        return maxOrbit;
+	    }
+	}
+	
+	public void computeVertexInvariants() {
+	    Map<Integer, int[]> layerInvariants = new HashMap<Integer, int[]>();
+	    for (int i = 0; i < this.nodes.size(); i++) {
+	        DAG.Node node = this.nodes.get(i);
+	        int j = node.vertexIndex;
+	        int[] layerInvariantsJ;
+	        if (layerInvariants.containsKey(j)) {
+	            layerInvariantsJ = layerInvariants.get(j);
+	        } else {
+	            layerInvariantsJ = new int[this.layers.size()];
+	            layerInvariants.put(j, layerInvariantsJ);
+	        }
+	        layerInvariantsJ[node.layer] = invariants.getNodeInvariant(i); 
+	    }
+	    
+	    List<InvariantArray> invariantLists = new ArrayList<InvariantArray>();
+	    for (int i : layerInvariants.keySet()) {
+	        InvariantArray invArr = new InvariantArray(layerInvariants.get(i), i); 
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: Iterating over [map.keySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#keySet--) and calling [map.get(Key)](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#get-java.lang.Object-) to get the values within the loop is inefficient. 
**Fix**: Iterate over [map.entrySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#entrySet--) and access the values using [entry.getValue()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.Entry.html#getValue--) to avoid potential performance penalty due to lookup.